hack/image-loader.sh

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env bash
+
+# Copyright 2019 The KubeOne Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# The image-loader script is used to pull all Docker images used by KubeOne,
+# Kubeadm, and Kubernetes, and push them to the the specified private registry.
+
+# WARNING: This script heavily depends on KubeOne and Kubernetes versions.
+# You must use the script coming the KubeOne release you've downloaded.
+
+set -euo pipefail
+
+KUBERNETES_VERSION=${KUBERNETES_VERSION:-}
+
+TARGET_REGISTRY=${TARGET_REGISTRY:-127.0.0.1:5000}
+PULL_OPTIONAL_IMAGES=${PULL_OPTIONAL_IMAGES:-true}
+
+# Wrapper around echo to include time
+function echodate() {
+  # do not use -Is to keep this compatible with macOS
+  echo "[$(date +%Y-%m-%dT%H:%M:%S%:z)]" "$@"
+}
+
+function fail() {
+  echodate "$@"
+  exit 1
+}
+
+function retag() {
+  local image="$1"
+
+  # Trim registry
+  local local_image
+  local name
+  local tag
+
+  local_image="$(echo "${image}" | cut -d/ -f1 --complement)"
+  # Split into name and tag
+  name="$(echo "${local_image}" | cut -d: -f1)"
+  tag="$(echo "${local_image}" | cut -d: -f2)"
+
+  # Build target image name
+  local target_image="${TARGET_REGISTRY}/${name}:${tag}"
+
+  echodate "Retagging \"${image}\" => \"${target_image}\"..."
+
+  docker pull "${image}"
+  docker tag "${image}" "${target_image}"
+  docker push "${target_image}"
+
+  echodate "Done retagging \"${image}\"."
+}
+
+# The script is only supported on Linux because it depends on Kubeadm.
+# You can run this script in a Docker container.
+if [[ "$OSTYPE" != "linux-gnu"* ]]; then
```

Comment:
This will error if the variable is not defined, no?

Reply:
Oh nevermind, this seems to be predefined somewhere. I only ever saw people explicitly calling uname to get the OS.